File: examples/java/com/intel/daal/examples/datasource/DataStructuresAOS.java
/* file: DataStructuresAOS.java */
/*******************************************************************************
* Copyright 2014-2018 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
 //  Content:
 //     Java example of using an array of structures (AOS)
 ////////////////////////////////////////////////////////////////////////////////
 */

/**
 * <a name="DAAL-EXAMPLE-JAVA-DATASTRUCTURESAOS">
 * @example DataStructuresAOS.java
 */

package com.intel.daal.examples.datasource;

import java.nio.DoubleBuffer;
import java.nio.IntBuffer;
import com.intel.daal.algorithms.cosdistance.*;
import com.intel.daal.data_management.data.AOSNumericTable;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.examples.utils.Service;
import java.io.Serializable;
import com.intel.daal.services.DaalContext;

class DataStructuresAOS {
    private static final int nVectors = 5;
    private static final long nFeatures = 4;
    private static final long firstReadRow = 0;

    public static class PointType implements Serializable {
        public float x;
        public float y;
        public int    categ;
        public double  value;

        public PointType(float _x, float _y, int _categ, double _value) {
            x = _x;
            y = _y;
            categ = _categ;
            value = _value;
        }
    }

    private static DaalContext context = new DaalContext();

    public static void main(String[] args) throws java.io.FileNotFoundException, java.io.IOException {

        System.out.println("Array of structures (AOS) numeric table example\n");
        PointType points[] = new PointType[nVectors];
        points[0] = new PointType(0.5f, -1.3f, 1, 100.1);
        points[1] = new PointType(2.5f, -3.3f, 2, 200.0);
        points[2] = new PointType(4.5f, -5.3f, 2, 350.3);
        points[3] = new PointType(6.5f, -7.3f, 0, 470.4);
        points[4] = new PointType(8.5f, -9.3f, 1, 270.5);

        /*
         * Construct AOS numericTable for a data array with nFeaturesAOS fields
         * and nPoints elements
         */
        AOSNumericTable dataTable = new AOSNumericTable(context, points);

        /* Create an algorithm to compute a cosine distance matrix using the defaultDense method */
        Batch alg = new Batch(context, Float.class, Method.defaultDense);

        alg.input.set(InputId.data, dataTable);
        Result result = alg.compute();

        NumericTable res = result.get(ResultId.cosineDistance);

        /* Read a block of rows */
        DoubleBuffer dataDouble = DoubleBuffer.allocate((int) (nVectors * nFeatures));
        dataDouble = res.getBlockOfRows(firstReadRow, nVectors, dataDouble);
        printDoubleBuffer(dataDouble, nFeatures, nVectors, "Print AOS data structures as double:");
        res.releaseBlockOfRows(firstReadRow, nVectors, dataDouble);

        int readFeatureIdx = 2;
        IntBuffer dataInt = IntBuffer.allocate((int) (nVectors * nFeatures));
        dataInt = res.getBlockOfColumnValues(readFeatureIdx, firstReadRow, nVectors, dataInt);
        printIntBuffer(dataInt, 1, nVectors, "Print the third feature of AOS:");
        res.releaseBlockOfRows(firstReadRow, nVectors, dataInt);


        context.dispose();
    }

    private static void printDoubleBuffer(DoubleBuffer buf, long nColumns, long nRows, String message) {
        int step = (int) nColumns;
        System.out.println(message);
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nColumns; j++) {
                System.out.format("%6.3f   ", buf.get(i * step + j));
            }
            System.out.println("");
        }
        System.out.println("");
    }

        private static void printIntBuffer(IntBuffer buf, long nColumns, long nRows, String message) {
        int step = (int) nColumns;
        System.out.println(message);
        for (int i = 0; i < nRows; i++) {
            for (int j = 0; j < nColumns; j++) {
                System.out.format("%3d   ", buf.get(i * step + j));
            }
            System.out.println("");
        }
        System.out.println("");
    }
}

File: examples/daal/cpp/source/datasource/datastructures_aos.cpp
/* file: datastructures_aos.cpp */
/*******************************************************************************
* Copyright 2014 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of using an array of structures (AOS)
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-DATASTRUCTURES_AOS">
 * @example datastructures_aos.cpp
 */

#include "daal.h"
#include "service.h"

using namespace daal;
using namespace daal::data_management;

struct PointType {
    float x;
    float y;
    int categ;
    double value;
};

int main() {
    std::cout << "Array of structures (AOS) numeric table example" << std::endl << std::endl;

    const size_t nObservations = 5;
    const size_t nFeatures = 4;
    PointType points[nObservations] = { { 0.5f, -1.3f, 1, 100.1 },
                                        { 2.5f, -3.3f, 2, 200.2 },
                                        { 4.5f, -5.3f, 2, 350.3 },
                                        { 6.5f, -7.3f, 0, 470.4 },
                                        { 8.5f, -9.3f, 1, 270.5 } };

    /* Construct AOS numericTable for a data array with nFeatures fields and nObservations elements*/
    AOSNumericTablePtr dataTable = AOSNumericTable::create(points, nFeatures, nObservations);
    checkPtr(dataTable.get());

    /* Add data to the numeric table */
    dataTable->setFeature<float>(0,
                                 DAAL_STRUCT_MEMBER_OFFSET(PointType, x),
                                 data_feature_utils::DAAL_CONTINUOUS);
    dataTable->setFeature<float>(1,
                                 DAAL_STRUCT_MEMBER_OFFSET(PointType, y),
                                 data_feature_utils::DAAL_CONTINUOUS);
    dataTable->setFeature<int>(2,
                               DAAL_STRUCT_MEMBER_OFFSET(PointType, categ),
                               data_feature_utils::DAAL_CATEGORICAL,
                               4);
    dataTable->setFeature<double>(3,
                                  DAAL_STRUCT_MEMBER_OFFSET(PointType, value),
                                  data_feature_utils::DAAL_CONTINUOUS);

    /* Read a block of rows */
    const size_t firstReadRow = 0;

    BlockDescriptor<double> doubleBlock;
    dataTable->getBlockOfRows(firstReadRow, nObservations, readOnly, doubleBlock);
    printArray<double>(doubleBlock.getBlockPtr(),
                       nFeatures,
                       doubleBlock.getNumberOfRows(),
                       "Print AOS data structures as double:");
    dataTable->releaseBlockOfRows(doubleBlock);

    /* Read a feature (column) */
    size_t readFeatureIdx = 2;

    BlockDescriptor<int> intBlock;
    dataTable->getBlockOfColumnValues(readFeatureIdx,
                                      firstReadRow,
                                      nObservations,
                                      readOnly,
                                      intBlock);
    printArray<int>(intBlock.getBlockPtr(),
                    1,
                    intBlock.getNumberOfRows(),
                    "Print the third feature of AOS:");
    dataTable->releaseBlockOfColumnValues(intBlock);

    return 0;
}
